### PR TITLE
Issue 3698

### PR DESCRIPTION
**Pull request by bmeike**
Change the order of args for the custom method to match the iOS call
Add an overloaded 2 arg usernamePassword method that defaults createUser false
Add unit tests as appropriate

### CODE REVIEW

File: realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java [hunk 1]
```diff
@@ -66,13 +66,39 @@
...
+    private final String identityProvider;
+    private final Map<String, Object> userInfo;
 
     // Factory constructors
 
     /**
+     * Creates credentials based on a Facebook login.
+     *
+     * @param facebookToken a facebook userIdentifier acquired by logging into Facebook.
+     * @return a set of credentials that can be used to log into the Object Server using
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: `.` at the end

File: realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java [hunk 2]
```diff
@@ -66,13 +66,39 @@
...
+        assertStringNotEmpty(facebookToken, "facebookToken");
+        return new SyncCredentials(facebookToken, IdentityProvider.FACEBOOK, null);
+    }
+
+    /**
+     * Creates credentials based on a Google login.
+     *
+     * @param googleToken a google userIdentifier acquired by logging into Google.
+     * @return a set of credentials that can be used to log into the Object Server using
+     *         {@link SyncUser#loginAsync(SyncCredentials, String, SyncUser.Callback)}
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: `.` at the end.

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/CredentialsTests.java [hunk 3]
```diff
@@ -98,24 +133,19 @@ public void usernamePassword_invalidUserName() {
...
-            try {
-                SyncCredentials.custom("facebook", username, userInfo);
-                fail();
-            } catch (IllegalArgumentException ignored) {
-            }
-        }
+    public void usernamePassword_nullPassword() {
+        SyncCredentials creds = SyncCredentials.usernamePassword("foo", null, true);
+        Map<String, Object> userInfo = creds.getUserInfo();
+        assertEquals(null, userInfo.get("password"));
```
Comment (beeender) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/CredentialsTests.java]: why not using `assertUsernamePassword` here?

File: realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java [hunk 4]
```diff
@@ -86,72 +112,56 @@
...
      * @throws IllegalArgumentException if any parameter is either {@code null} or empty.
      */
-    public static SyncCredentials custom(String identityProvider, String userIdentifier, Map<String, Object> userInfo) {
-        if (identityProvider == null || identityProvider.equals("")) {
-            throw new IllegalArgumentException("Non-null 'identityProvider' required.");
-        }
-        if (userIdentifier == null || userIdentifier.equals("")) {
-            throw new IllegalArgumentException("Non-null 'userIdentifier' required.");
-        }
+    public static SyncCredentials custom(String userIdentifier, String identityProvider, Map<String, Object> userInfo) {
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: We need to add a changelog item to the Object Server section about this, as it is a breaking semantic change.

File: realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java (hunk 1 above)
Comment (bmeike) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: Did you guys leave these problems here just to haze me??? ;-P

Comment (bmeike) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: Fixed

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/CredentialsTests.java [hunk 5]
```diff
@@ -27,6 +27,7 @@
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
+
```
Comment (bmeike) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/CredentialsTests.java]: Fixed.  I would like to point out that his stuff is making it past both checkstyle and lint!!

File: realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java (hunk 4 above)
Comment (bmeike) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: Right.  I wondered about that.  Fixed.

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/CredentialsTests.java (hunk 3 above)
Comment (bmeike) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/CredentialsTests.java]: Good call!  Fixed!

**Comment by bmeike** [comment]
All comments addressed.

File: realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java (hunk 1 above)
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/SyncCredentials.java]: Yes! We did!! We put traps everywhere!

**Pull request merged**